lib/compilers/d8.ts

```diff
@@ -0,0 +1,341 @@
+        const d8Options = [
+            '-cp',
+            this.compiler.exe, // R8 jar.
+            'com.android.tools.r8.D8', // Main class name for the D8 compiler.
+            ...options.slice(0, sourceFileOptionIndex),
+            ...classFiles,
+        ];
+        const result = await this.exec(this.javaPath, d8Options, execOptions);
+        return {
+            ...this.transformToCompilationResult(result, outputFilename),
+            languageId: this.getCompilerResultLanguageId(),
+        };
+    }
+
+    override async objdump(outputFilename, result: any, maxSize: number) {
+        const dirPath = path.dirname(outputFilename);
+
+        // There is only one dex file for all classes.
+        let files = await fs.readdir(dirPath);
+        const dexFile = files.find(f => f.endsWith('.dex'));
+        const baksmaliOptions = ['-jar', this.compiler.objdumper, 'd', `${dexFile}`, '-o', dirPath];
+        const baksmaliResult = await this.exec(this.javaPath, baksmaliOptions, {
+            maxOutput: maxSize,
+            customCwd: dirPath,
+        });
+
+        // There is one smali file for each class.
+        files = await fs.readdir(dirPath);
+        const smaliFiles = files.filter(f => f.endsWith('.smali'));
+        let objResult = '';
+        for (const smaliFile of smaliFiles) {
+            objResult = objResult.concat(fs.readFileSync(path.join(dirPath, smaliFile), 'utf-8') + '\n\n');
+        }
+
+        const asmResult: ParsedAsmResult = {
+            asm: [
+                {
+                    text: objResult,
+                },
+            ],
+        };
+
+        result.asm = asmResult.asm;
+        return result;
+    }
+
+    override optionsForFilter(filters: ParseFiltersAndOutputOptions) {
+        filters.binary = true;
+        return [];
+    }
+
+    // Map line numbers to lines.
+    override async processAsm(result) {
+        const asm = result.asm[0].text;
+        const segments: ParsedAsmResultLine[] = [];
+
+        let lineNumber;
+        for (const l of asm.split(/\n/)) {
+            if (this.lineNumberRegex.test(l)) {
+                lineNumber = Number.parseInt(l.match(this.lineNumberRegex)[1]);
+                segments.push({text: l, source: null});
+            } else if (this.methodEndRegex.test(l)) {
+                lineNumber = null;
+                segments.push({text: l, source: null});
+            } else {
+                if (/\S/.test(l) && lineNumber) {
+                    segments.push({text: l, source: {file: null, line: lineNumber}});
+                } else {
+                    segments.push({text: l, source: null});
+                }
+            }
+        }
+        return {asm: segments};
+    }
+
+    // This is messy, but a PreliminaryCompilerInfo needs to be created for the
+    // JavaCompiler/KotlinCompiler instances.
+    static createJavaCompilerInfo(javacPath: string): PreliminaryCompilerInfo {
```

Comment:
This is indeed super messy and quite fragile too. This is so you can run the java compiler? I'm sure there must be a better way to do this... I thought there must be a way to get the _actual_ compiler object but seems not. I wonder if @partouf or @jeremy-rifkin have a cunning idea here?

Reply:
This information is all supposed to come from the compiler-finder (that's the point of `PreliminaryCompilerInfo`, it represents basic information collected by the finder and then the base compiler fills in the omitted properties of `CompilerInfo`). There should be a way to get an actual compiler object and not have to construct a new one every time `runCompiler` is called, but looking through the code I actually don't see one, at least right off hand.

`CompileHandler` has a `findCompiler(langId, compilerId)` and `compilersById: Record<string, Record<string, BaseCompiler>>` but I didn't see any other only place constructed compilers are saved. I think we might want to pull something out globally for this.

Reply:
> I think we might want to pull something out globally for this.

Just to make sure I'm understanding correctly, do you mean updating `CompileHandler` so that either `findCompiler(langId, compilerId)` or `compilersById: Record<string, Record<string, BaseCompiler>>` can be accessed from `d8.ts`?

Reply:
Hi @kevinjeon-g, more or less yes. It's not entirely clear to me the best way to go about this, perhaps making the `handlerConfig` in app.ts (which has the `CompileHandler` instance) an exported global could work. I'm happy to help with this since it involves some bigger picture design decisions in the codebase, or if you'd like to tackle it yourself go for it :)

Reply:
Thanks @jeremy-rifkin ! Making `handlerConfig` global was a pretty straightforward option, and it seems to work as expected here. This allowed me to remove the functions for making a fake compiler config, and should also avoid creating compilers altogether.

I think this should be safe as long as the properties for android-java/android-kotlin (java/kotlin compiler IDs and binary paths) are configured correctly.